simde/x86/avx512/ror.h

```diff
@@ -0,0 +1,276 @@
+#if !defined(SIMDE_X86_AVX512_ROR_H)
+#define SIMDE_X86_AVX512_ROR_H
+
+#include "types.h"
+#include "mov.h"
+#include "or.h"
+#include "srli.h"
+#include "slli.h"
+#include "../avx2.h"
+
+HEDLEY_DIAGNOSTIC_PUSH
+SIMDE_DISABLE_UNWANTED_DIAGNOSTICS
+SIMDE_BEGIN_DECLS_
+
+#if defined(SIMDE_X86_AVX512F_NATIVE) && defined(SIMDE_X86_AVX512VL_NATIVE)
+  #define simde_mm_ror_epi32(a, imm8) _mm_ror_epi32(a, imm8)
+#else
+  SIMDE_FUNCTION_ATTRIBUTES
+  simde__m128i
+  simde_mm_ror_epi32_internal_(simde__m128i a, simde__m128i result, int imm8)
+      SIMDE_REQUIRE_CONSTANT_RANGE (imm8, 0, 255) {
+    simde__m128i retval;
+
+    if(imm8 & 31)
+      retval = result;
+    else
+      retval = a;
```

Comment:
For cases like this it's generally a good idea to get into the habit of using the ternary operator.  In this case, the whole function could just be

```c
return (imm8 & 31) ? result : a;
```

Not only does that make the code a bit easier to read IMHO (as long as you understand the ternary operator, which I think is a reasonable expectation for people reading the SIMDe source code), it also eliminates the if / else.  When writing performance-sensitive code (especially when we want the compiler to vectorize) if/else is a bit of a code smell; in this case I'm sure any compiler will handle it exactly the same, but if you were doing something a bit more complicated like:

```c
if (imm8 & 31)
  foo = result;
else
  bar = result;
```

That will cause a branch in the compiled code, and branches are *slow*.  The ternary operator is really nice because you can't really do this sort of general-purpose branching; all you can do is "based on condition, the result is either X or Y", which is generally handled much better in hardware.  On the scalar side, you have the `cmov` instruction on x86, and in SIMD you have various bitselect/blend functions (see `wasm_v128_bitselect` in WASM, it has a bunch of implementations on other architectures).

Reply:
I agree, that is what I tried originally but ran into that bug with icc we talked about.

Reply:
That's the issue with ICC? I thought the icc bug was the `SIMDE_VECTORIZE_` on the loop.

That's rather disconcerting, actually.  That should definitely go in an ifdef, then, otherwise someone is likely to come along and "fix" it to be just `return (imm8 & 31) ? result : a;`  It sounds like this is going to have to go away anyways because of the macro issue above, though…